### PR TITLE
Use the Hungarian algorithm for maximum-weight matching

### PR DESCRIPTION
**Pull request by dourouc05**
Uses a bridge to Hungarian.jl. Follows #6.

### CODE REVIEW

**Comment by dourouc05** [comment]
I am working on the failure…

**Comment by matbesancon** [comment]
Nice work!
It seems that the Hungarian.jl package does not accept AbstractMatrix but concrete Matrix for one function you're calling `munkres`. Maybe make them a PR and adapt the test while we wait for them to update it, or change the function here to Matrix (which is a bit sad).

**Comment by matbesancon** [comment]
Also, I know this is a bother but don't use the try catch around `import BlossomV`, this blocks precompilation, see https://github.com/JuliaGraphs/LightGraphs.jl/pull/815 for reference, Julio mentioned it.

**Comment by dourouc05** [comment]
Just for the follow-up: I made to PRs to this package, the first one to solve the error messages (https://github.com/Gnimuc/Hungarian.jl/pull/7), the second to handle edges that are not in the graph (https://github.com/Gnimuc/Hungarian.jl/pull/8). I am waiting for the second to get merged before finishing this.

**Comment by matbesancon** [comment]
Perfect, let me know when the PRs are submitted, I'll re-run the CI checks

**Comment by dourouc05** [comment]
Both PRs are merged, but I still have a few things to get right on LGMatching's side.

**Comment by dourouc05** [comment]
Here it is!

**Comment by matbesancon** [comment]
Hungarian.jl build still seem to fail or in progress, we'll have to wait and see if any further bugs occur before merging this.

**Comment by matbesancon** [comment]
Also what we do here is add a new dependency for the same function as something already implemented. This could be more appropriate to have a unique `maximum_weight_maximal_matching` function which takes an algorithm keyword argument which would be one of MIP or Hungarian.
See LightGraphsFlows.jl

**Comment by dourouc05** [comment]
I thought that would be done in a second step, but I'll look into this! (Just defining a class per algorithm, and make dispatch work properly, it seems.)

**Comment by matbesancon** [comment]
That's it, also define an abstract type MaxWeightMatchingAlgorithm from
which the two algorithms inherit ;)

**Comment by matbesancon** [comment]
@dourouc05 some news on that? Do you need help fixing the errors?

**Comment by Gnimuc** [comment]
Is there anything I can help with?

**Comment by matbesancon** [comment]
yes, we need to rebase this branch on master to see if all is well, if you have time it would be great :)

**Comment by dourouc05** [comment]
Yes, I'm still here :)! However, for rebasing, I don't think I will be able to make it soon: I'm still stuck on 0.6 due to JuMP (or maybe 0.7, but definitely not 1.0) -- actually, JuMP is the main reason I'm using Julia. 

I'm first going through the implementation again, then add `MaxWeightMatchingAlgorithm`, and I'll be back here!

**Comment by matbesancon** [comment]
JuMP and MPB are available on Julia 1.0 now, you can rebase on master without fear

File: REQUIRE
Comment (matbesancon) [inline REQUIRE]: missing new line at end of file

File: REQUIRE
Comment (matbesancon) [inline REQUIRE]: we shouldn't depend on patches, only versions x.y, not x.y.z

File: src/hungarian.jl
Comment (matbesancon) [inline src/hungarian.jl]: `for i in eachindex(assignment)`

File: src/hungarian.jl
Comment (matbesancon) [inline src/hungarian.jl]:
```julia
cost = sum(w[p[1], p[2]] for p in pairs)
```
No need to loop

**Review by matbesancon — CHANGES_REQUESTED** [review]

File: REQUIRE
Comment (dourouc05) [reply REQUIRE]: The thing is that a specific functionality was added in this version (handling `missing` values).

File: REQUIRE
Comment (Gnimuc) [inline REQUIRE]: it would be nice to depend on Hungarian 0.2 which includes additional features and performance enhancement. 
```suggestion
Hungarian 0.2
```

File: REQUIRE
Comment (matbesancon) [reply REQUIRE]: ok, if we use version x.y this should fetch x.y.z with the biggest z
Otherwise as mentioned Hungarian 0.2 will do the job, also we can update JuMP to v0.18 I think

**Comment by dourouc05** [comment]
Here it is! I also added a new interface for the two algorithms solving the same problem, as proposed by @matbesancon. The tests pass locally, on Julia 0.6.

**Comment by matbesancon** [comment]
ok wonderful, can you check it with Julia 1.0? This also needs conflict fixing with master.
Some help: on Julia 1.0
```
julia> ]
pkg> add LightGraphsMatching
pkg> dev LightGraphsMatching
```

Then in `.julia/dev/LightGraphsMatching`, pull your branch, merge master, test and re-push to the branch

**Comment by dourouc05** [comment]
I am trying to test on the new Julia, but installing Cbc is causing me troubles (impossible to install libstdc++, `The system cannot find the file specified.`). Here is the merged tree, though (it compiles).

**Review by matbesancon — APPROVED** [review]

**Comment by matbesancon** [comment]
there was an error during tests (some variable U undefined or so:
https://travis-ci.org/JuliaGraphs/LightGraphsMatching.jl/jobs/451089016

**Comment by dourouc05** [comment]
It should be solved with the latest commit.

File: src/hungarian.jl
Comment (matbesancon) [inline src/hungarian.jl]: There was an error here, this needs `wDual = maximum(w) .- w`

**Review by matbesancon — CHANGES_REQUESTED** [review]

**Comment by dourouc05** [comment]
I just solved deprecations (Julia 0.6 syntax).

**Comment by matbesancon** [comment]
all good, thanks a lot @dourouc05 for this work!

**Pull request merged**
